dev/com.ibm.ws.threading_policy_fat/test-applications/basicfat/src/web/PolicyExecutorServlet.java

```diff
@@ -2397,6 +2397,183 @@ public void testInvokeAllTimeoutWaitForEnqueue() throws Exception {
         assertEquals(0, canceledFromQueue.size());
     }
 
+    // Basic test of timed invokeAny. Submit a group of tasks, of which one raises an exception,
+    // another runs for longer than the timeout, and another completes successfully.
+    // The result of the one that completed successfully should be returned,
+    // and the duration of the invokeAny method should not exceed the timeout.
+    @Test
+    public void testInvokeAnyTimed1Successful() throws Exception {
+        PolicyExecutor executor = provider.create("testInvokeAnyTimed1Successful");
+
+        CountDownLatch beginLatchForBlocker = new CountDownLatch(1);
+        CountDownLatch blocker = new CountDownLatch(1);
+        CountDownLatch unused = new CountDownLatch(0);
+
+        List<Callable<Boolean>> tasks = new ArrayList<Callable<Boolean>>();
+        tasks.add(new CountDownTask(unused, null, 0)); // intentionally cause NullPointerException
+        tasks.add(new CountDownTask(beginLatchForBlocker, blocker, TIMEOUT_NS * 2)); // block for longer than the timeout
+        tasks.add(new CountDownTask(unused, beginLatchForBlocker, TIMEOUT_NS / 2)); // should succeed
+
+        long start = System.nanoTime();
+        assertTrue(executor.invokeAny(tasks, TIMEOUT_NS, TimeUnit.NANOSECONDS));
+        long duration = System.nanoTime() - start;
+
+        assertTrue(duration + "ns", duration < TIMEOUT_NS);
+
+        List<Runnable> canceledFromQueue = executor.shutdownNow();
+        assertEquals(0, canceledFromQueue.size());
+    }
+
+    // Submit a group of tasks to invokeAny, of which one raises an exception and
+    // another runs for longer than the timeout. TimeoutException should occur.
+    @Test
+    public void testInvokeAnyTimed1TimesOut1Fails() throws Exception {
+        PolicyExecutor executor = provider.create("testInvokeAnyTimed1TimesOut1Fails");
+
+        CountDownLatch blocker = new CountDownLatch(1);
+        CountDownLatch unused = new CountDownLatch(0);
+
+        List<Callable<Boolean>> tasks = new ArrayList<Callable<Boolean>>();
+        tasks.add(new CountDownTask(unused, null, 0)); // intentionally cause NullPointerException
+        tasks.add(new CountDownTask(unused, blocker, TIMEOUT_NS * 2)); // block for longer than the timeout
+
+        long start = System.nanoTime();
+        try {
+            fail("Should not return value: " + executor.invokeAny(tasks, 200, TimeUnit.MILLISECONDS));
+        } catch (TimeoutException x) { // pass
+        }
+        long duration = System.nanoTime() - start;
+
+        assertTrue(duration + "ns", duration < TIMEOUT_NS); // higher value used to allow for slowness and pauses in test systems
```

Comment:
If the timeout is 200 ms, should we be checking the duration for something less than TIMEOUT_NS, which is 2 minutes?

Reply:
No, because TIMEOUT_NS is defined as the minimal amount of tolerance we need to allow when expecting anything to happen in order to avoid being vulnerable to random pauses and slowness in test infrastructure.  Given that, the best we can do is verify that the invokeAny operation does indeed time out rather than blocking until when the task would have otherwise completed.